### DISCUSSION TITLE
Best practice for a sliding door as Kinematic body ?

### DISCUSSION THREAD

**Discussion by Froyok** [jrouwe/JoltPhysics#1302] [Q&A] [answered]
I wanted to create a simple sliding door (the kind you see at the entrance of your local supermarket). 
To me it seemed like the best approach for this would be to make a kinematic body and drive its movement via code to have control over its animation speed (with speedup and slow down, or something complex).

I looked at the [following discussion](https://github.com/jrouwe/JoltPhysics/discussions/178) which explained that to properly move the body I needed to use `MoveKinematic()` instead of `SetPosition()`/`SetRotation()`. That works, but what surprised me is that once the system update is done, the body velocity does not go back to 0 even when `MoveKinematic()` isn't called anymore. I have to manually set it to 0 to ensure the body stops moving.

This is roughly how I move the object in my code:

```lua
-- Called once per frame while the door is moving
function Door.Update( Deltatime )
	if Animation.Time < Animation.TimeMax then
		Animation.Time = Animation.Time + Deltatime

		-- Animation progress from 0 to 1
		TimeElapsed = Animation.Time / Animation.TimeMax
		TimeElapsed = math.clamp( TimeElapsed, 0.0, 1.0 )

		-- Inverse animation to close door if already open
		if IsOpen then
			TimeElapsed = 1.0 - TimeElapsed
		end

		TempVec1 = Position + DirectionAxis * Distance
		TempVec1 = Lerp( Position, TempVec1, TimeElapsed )

		BodyMoveKinematic(
			Body,
			TempVec1,
			Rotation,
			math.max( ENGINE_DELTATIME, 0.001 )
		)
	end
end
```

So basically I use `MoveKinematic()` the way I would use `SetPosition()`.
I was wondering if that was the expected way of managing this kind of objects ? 
If so, is killing the velocity the expected outcome afterward ?

**Comment by jrouwe** [comment, marked as answer]
`MoveKinematic` will calculate the velocity needed to go from where you are to the desired position in the desired delta time (= usually a single physics update). The physics engine will never change the velocity of a kinematic body, you need to do this yourself. If you stop calling `MoveKinematic` the body will continue moving at the same speed forever. If you want the body to stop after arriving then call `BodyInterface::SetLinearAndAngularVelocity(body_id, Vec3::sZero(), Vec3::sZero())`.

**Reply by Froyok** [reply]
Sounds good, thank you for your answer ! :)
